Source/Initialization/WarpXInitData.cpp

```diff
@@ -231,23 +233,228 @@ WarpX::PostRestart ()
+ *  initialization assumes that the function has three independent
+ *  variables, at max, namely, x, y, z. However, any number of constants
+ *  can be used in the function used to define the E and B fields on the grid.
+ */
+
 void
 WarpX::InitLevelData (int lev, Real time)
 {
+
+    ParmParse pp("warpx");
+
+    // default values of E_external_grid and B_external_grid
+    // are used to set the E and B field when "constant" or
+    // "parser" is not explicitly used in the input.
+    pp.query("B_ext_grid_init_style", B_ext_grid_s);
+    std::transform(B_ext_grid_s.begin(),
+                   B_ext_grid_s.end(),
+                   B_ext_grid_s.begin(),
+                   ::tolower);
+
+    pp.query("E_ext_grid_init_style", E_ext_grid_s);
+    std::transform(E_ext_grid_s.begin(),
+                   E_ext_grid_s.end(),
+                   E_ext_grid_s.begin(),
+                   ::tolower);
+
+    // if the input string is "constant", the values for the
+    // external grid must be provided in the input.
+    if (B_ext_grid_s == "constant")
+        pp.getarr("B_external_grid", B_external_grid);
+
+    // if the input string is "constant", the values for the
+    // external grid must be provided in the input.
+    if (E_ext_grid_s == "constant")
+        pp.getarr("E_external_grid", E_external_grid);
+
     for (int i = 0; i < 3; ++i) {
         current_fp[lev][i]->setVal(0.0);
-        Efield_fp[lev][i]->setVal(E_external_grid[i]);
-        Bfield_fp[lev][i]->setVal(B_external_grid[i]);
+        if (B_ext_grid_s == "constant" || B_ext_grid_s == "default")
+           Bfield_fp[lev][i]->setVal(B_external_grid[i]);
+        if (E_ext_grid_s == "constant" || E_ext_grid_s == "default")
+           Efield_fp[lev][i]->setVal(E_external_grid[i]);
+    }
+
+    if (B_ext_grid_s == "parse_b_ext_grid_function") {
+
+       std::vector<std::string> f;
+       // Parse Bx_external_grid_function
+       pp.getarr("Bx_external_grid_function(x,y,z)", f);
+       str_Bx_ext_grid_function.clear();
+       for (auto const& s : f) {
+           str_Bx_ext_grid_function += s;
+       }
+       f.clear();
+
+       // Parse By_external_grid_function
+       pp.getarr("By_external_grid_function(x,y,z)", f);
+       str_By_ext_grid_function.clear();
+       for (auto const& s : f) {
+            str_By_ext_grid_function += s;
+       }
+       f.clear();
+
+       // Parse Bz_external_grid_function
+       pp.getarr("Bz_external_grid_function(x,y,z)", f);
+       str_Bz_ext_grid_function.clear();
+       for (auto const& s : f) {
+           str_Bz_ext_grid_function += s;
+       }
+       f.clear();
+
+       // Initialize Bfield_fp with external function
+       MultiFab *Bx, *By, *Bz;
+       Bx = Bfield_fp[lev][0].get();
+       By = Bfield_fp[lev][1].get();
+       Bz = Bfield_fp[lev][2].get();
+
+       bool B_flag = 1;
+       InitializeExternalFieldsOnGridUsingParser(Bx, By, Bz, lev, B_flag);
+    }
```

Comment:
The piece of code above does 4 times the same thing for the magnetic field, and the piece of code below does the same for electric field. Could we avoid this duplication?

Reply:
```c++
if (B_ext_grid_s == "parse_b_ext_grid_function") {
    call_a_function("Bx_external_grid_function(x,y,z)", str_Bx_ext_grid_function);
    call_a_function("By_external_grid_function(x,y,z)", str_By_ext_grid_function);
    call_a_function("Bz_external_grid_function(x,y,z)", str_Bz_ext_grid_function);
}
?
```
and same for E?

Reply:
Thank you @MaxThevenet . I have added the suggested improvement to the PR now.

Reply:
As a follow-up from our conversation, I have removed B_flag=0 or 1, and instead pass the nodal_flag IntVects and parser_strings to the InitializeExternal...() function